pkg/cmd/roachtest/cdc.go

```diff
@@ -157,6 +159,156 @@ func cdcBasicTest(ctx context.Context, t *test, c *cluster, args cdcTestArgs) {
+	c.Put(ctx, cockroach, "./cockroach", crdbNodes)
+	c.Put(ctx, workload, "./workload", workloadNode)
+	c.Start(ctx, t, crdbNodes)
+	kafka := kafkaManager{
+		c:     c,
+		nodes: kafkaNode,
+	}
+	kafka.install(ctx)
+	kafka.start(ctx)
+	defer kafka.stop(ctx)
+
+	c.Run(ctx, workloadNode, `./workload init bank {pgurl:0}`)
+	db := c.Conn(ctx, 1)
+	defer stopFeeds(db)
+
+	if _, err := db.Exec(
+		`SET CLUSTER SETTING changefeed.experimental_poll_interval = '0ns'`,
+	); err != nil {
+		t.Fatal(err)
+	}
+	var jobID string
+	if err := db.QueryRow(
+		`CREATE CHANGEFEED FOR bank.bank INTO $1 WITH updated, resolved`, kafka.sinkURL(ctx),
+	).Scan(&jobID); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Status("running workload")
+	workloadCtx, workloadCancel := context.WithCancel(ctx)
+	m := newMonitor(workloadCtx, c, crdbNodes)
+	var doneAtomic int64
+	m.Go(func(ctx context.Context) error {
+		err := c.RunE(ctx, workloadNode, `./workload run bank {pgurl:0}`)
+		if atomic.LoadInt64(&doneAtomic) > 0 {
+			return nil
+		}
+		return err
+	})
+	m.Go(func(ctx context.Context) error {
+		defer workloadCancel()
+		l, err := c.l.ChildLogger(`changefeed`)
+		if err != nil {
+			return err
+		}
+		defer l.close()
+
+		tc, err := kafka.consumer(ctx, `bank`)
+		if err != nil {
+			return err
+		}
+		defer tc.Close()
+
+		// TODO(dan): Force multiple partitions and re-enable this check. It
+		// used to be the only thing that verified our correctness with multiple
+		// partitions but TestValidations/enterprise also does that now, so this
+		// isn't a big deal.
+		//
+		// if len(tc.partitions) <= 1 {
+		//  return errors.New("test requires at least 2 partitions to be interesting")
+		// }
+
+		const requestedResolved = 100
+		var numResolved, rowsSinceResolved int
+		v := changefeedccl.Validators{
+			changefeedccl.NewOrderValidator(`bank`),
+			// TODO(mrtracy): Disabled by #30902. Re-enabling is tracked by #31110.
+			// changefeedccl.NewFingerprintValidator(sqlDB.DB, `bank`, `fprint`, tc.partitions),
+		}
+		if _, err := db.Exec(
+			`CREATE TABLE fprint (id INT PRIMARY KEY, balance INT, payload STRING)`,
+		); err != nil {
+			return err
+		}
+
+		for {
+			m := tc.Next(ctx)
+			updated, resolved, err := changefeedccl.ParseJSONValueTimestamps(m.Value)
+			if err != nil {
+				return err
+			}
+
+			partitionStr := strconv.Itoa(int(m.Partition))
+			if len(m.Key) > 0 {
+				v.NoteRow(partitionStr, string(m.Key), string(m.Value), updated)
+				rowsSinceResolved++
+			} else {
+				if err := v.NoteResolved(partitionStr, resolved); err != nil {
+					return err
+				}
+				if rowsSinceResolved > 0 {
+					numResolved++
+					if numResolved > requestedResolved {
+						atomic.StoreInt64(&doneAtomic, 1)
+						break
+					}
+				}
+				rowsSinceResolved = 0
+			}
+		}
+		if failures := v.Failures(); len(failures) > 0 {
+			return errors.New(strings.Join(failures, "\n"))
+		}
+		return nil
+	})
+	m.Wait()
+}
+
+func runCDCErrors(ctx context.Context, t *test, c *cluster) {
```

Comment:
This doesn't need to be a roachtest at all; none of these cases even contact kafka!